### PR TITLE
feat: adding kafka

### PR DESCRIPTION
**Pull request by datascienceChris**
### Link to Github Issue
---
 
https://github.com/hipagesgroup/data-tools/issues/18

### What changes are proposed in this PR?
---
Adding wrappers to the kafka functionality of confluent kafka

### How was this tested?
---
Unit tests provided

### CODE REVIEW

File: hip_data_tools/apache/kafka.py [hunk 1]
```diff
@@ -0,0 +1,599 @@
...
+    def subscribe_consumer(self):
+        """
+        Subscribe the Kafka consumer to the given topic
+        Returns: None
+
+        """
+        self._kafka_consumer.subscribe([self._topic])
+        self._subscribed_to_topic = True
+
+    def get_msgs(self):
```
[the remark this answers is not in the document]
Comment (arihantsurana) [reply hip_data_tools/apache/kafka.py]: this seems like it needs to be fixed.

File: setup.py [hunk 2]
```diff
@@ -1,11 +1,33 @@
...
 
 from hip_data_tools.common import get_release_version, get_long_description
 
+
+class PyTest(test_command):
+    user_options = [('pytest-args=', 'a', "Arguments to pass to py.test")]
+
+    def initialize_options(self):
+        test_command.initialize_options(self)
+        self.pytest_args = ["-v"]
```
[the remark this answers is not in the document]
Comment (arihantsurana) [reply setup.py]: fix?

**Comment by arihantsurana** [comment]
yeah you got some legitimate code coverage issues and complexity issues.

File: hip_data_tools/apache/kafka.py [hunk 3]
```diff
@@ -0,0 +1,599 @@
...
+        root_path (str): path for data to be exported
+        bucket (str): Name of bucket for export
+        ts_col_nm (str): Name of the timestamp column used to partition the
+            messages
+        partition_key_nm (str): Name of the column in which to deposit the
+            rounded timestamps used to partition the data
+    Returns (KafkaS3BatchExporter): Instantiated Exporter
+
+    """
+    root_path = root_path if root_path is not None else os.environ.get(
```
Comment (arihantsurana) [inline hip_data_tools/apache/kafka.py]: perhaps read the env vars in a `KafkaConfig` class, and pass in that object here?

File: hip_data_tools/apache/kafka.py [hunk 4]
```diff
@@ -0,0 +1,599 @@
...
+            "KAFKA_PARTITION_KEY_NM")
+
+    aws_region = os.environ.get("AWS_DEFAULT_REGION")
+
+    conn = AwsConnection(mode="standard_env_var", region_name=aws_region,
+                         settings={})
+
+    s3_client = S3Util(conn, bucket)
+
+    return KafkaS3BatchExporter(root_path,
```
Comment (arihantsurana) [inline hip_data_tools/apache/kafka.py]: pass in kafka_config object

File: hip_data_tools/apache/kafka.py [hunk 5]
```diff
@@ -0,0 +1,694 @@
...
+            if bootstrap_servers is None else bootstrap_servers
+
+        self.configs = literal_eval(get_from_env_or_default_with_warning(
+            "KAFKA_PRODUCER_CONF", DEFAULT_PRODUCER_CONF)) if conf is None \
+            else conf
+
+        self.configs.update({'bootstrap.servers': self._bootstrap_servers})
+
+
+def create_producer(kafka_producer_config=None):
```
[the remark this answers is not in the document]
Comment (datascienceChris) [reply hip_data_tools/apache/kafka.py]: the similar blocks of code do different things

File: hip_data_tools/apache/kafka.py [hunk 6]
```diff
@@ -0,0 +1,705 @@
...
+from datetime import datetime
+
+import pandas as pd
+from confluent_kafka import Producer, Consumer
+
+from hip_data_tools.authenticate import AwsConnection
+from hip_data_tools.aws.s3 import S3Util
+from hip_data_tools.common import get_from_env_or_default_with_warning
+
+# TODO:
```
[the remark this answers is not in the document]
Comment (arihantsurana) [reply hip_data_tools/apache/kafka.py]: empty todo

File: setup.py [hunk 7]
```diff
@@ -1,11 +1,33 @@
...
 """
+import sys
 
 from setuptools import setup, find_packages
+from setuptools.command.test import test as test_command
 
 from hip_data_tools.common import get_release_version, get_long_description
 
+
+class PyTest(test_command):
```
[the remark this answers is not in the document]
Comment (arihantsurana) [reply setup.py]: fix?

**Comment by arihantsurana** [comment]
Make smaller PRs

**Pull request merged**